pkg/hostagent/snats.go

```diff
@@ -0,0 +1,476 @@
+			agent.snatLocalInfoUpdate(obj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			agent.snatLocalInfoDelete(obj)
+		},
+	})
+	agent.log.Debug("Initializing Snat Local info Informers")
+}
+
+func (agent *HostAgent) initSnatGlobalInformerBase(listWatch *cache.ListWatch) {
+	agent.snatGlobalInformer = cache.NewSharedIndexInformer(
+		listWatch,
+		&snatglobal.SnatGlobalInfo{},
+		controller.NoResyncPeriodFunc(),
+		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+	)
+	agent.snatGlobalInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			agent.snatGlobalInfoUpdate(obj)
+		},
+		UpdateFunc: func(_ interface{}, obj interface{}) {
+			agent.snatGlobalInfoUpdate(obj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			agent.snatGlobalInfoDelete(obj)
+		},
+	})
+	agent.log.Debug("Initializing SnatGlobal Info Informers")
+}
+
+func (agent *HostAgent) snatLocalInfoUpdate(obj interface{}) {
+	agent.indexMutex.Lock()
+	defer agent.indexMutex.Unlock()
+	snat := obj.(*snatlocal.SnatLocalInfo)
+	key, err := cache.MetaNamespaceKeyFunc(snat)
+	if err != nil {
+		SnatLocalInfoLogger(agent.log, snat).
+			Error("Could not create key:" + err.Error())
+		return
+	}
+	agent.log.Info("Snat Local Object added/Updated ", snat)
+	agent.doUpdateSnatLocalInfo(key)
+}
+
+func (agent *HostAgent) snatGlobalInfoUpdate(obj interface{}) {
+	agent.indexMutex.Lock()
+	defer agent.indexMutex.Unlock()
+	snat := obj.(*snatglobal.SnatGlobalInfo)
+	key, err := cache.MetaNamespaceKeyFunc(snat)
+	if err != nil {
+		SnatGlobalInfoLogger(agent.log, snat).
+			Error("Could not create key:" + err.Error())
+		return
+	}
+	agent.log.Info("Snat Local Object added/Updated ", snat)
+	agent.doUpdateSnatGlobalInfo(key)
+}
+
+func (agent *HostAgent) doUpdateSnatLocalInfo(key string) {
+	snatobj, exists, err :=
+		agent.snatLocalInformer.GetStore().GetByKey(key)
+	if err != nil {
+		agent.log.Error("Could not lookup snat for " +
+			key + ": " + err.Error())
+		return
+	}
+	if !exists || snatobj == nil {
+		return
+	}
+	snat := snatobj.(*snatlocal.SnatLocalInfo)
+	logger := SnatLocalInfoLogger(agent.log, snat)
+	agent.snatLocalInfoChanged(snatobj, logger)
+}
+
+func (agent *HostAgent) snatLocalInfoChanged(snatobj interface{}, logger *logrus.Entry) {
+	snat := snatobj.(*snatlocal.SnatLocalInfo)
+	if logger == nil {
+		logger = agent.log.WithFields(logrus.Fields{})
+	}
+	logger.Debug("Snat local info Changed...")
+	localInfo := snat.Spec.LocalInfos
+	syncep := false
+	// This case is true when scope moves from namespace to deployment
+	if len(localInfo) < len(agent.opflexSnatLocalInfos) {
+		for poduid, v := range agent.opflexSnatLocalInfos {
+			if _, ok := localInfo[poduid]; !ok {
+				// if pod present mark it snat ip deleted for local info
+				if _, ok := agent.opflexEps[poduid]; ok {
+					v.MarkDelete = true
+					syncep = true
+				} else {
+					delete(agent.opflexSnatLocalInfos, poduid)
```

Comment:
There are changes to the cache happening here and below, you don't need to acquire lock?

Reply:
calling function acquired the lock after that it is deferred the lock until the function execution completes